packages/framework/src/Model/Administrator/Security/AdministratorUserProvider.php

```diff
@@ -95,6 +103,12 @@ public function refreshUser(UserInterface $userInterface)
             $this->administratorActivityFacade->updateCurrentActivityLastActionTime($freshAdministrator);
         }
 
+        if ($freshAdministrator->getRolesChangeAt() > $administrator->getRolesChangeAt()) {
```

Comment:
I do not understand why you are using listener here... why not directly access token and change it?

Reply:
It is not possible to change token here. 
`AdministratorUserProvider::refreshUser` is called at the very beginning of the request, before firewall initialization (`security.yml`). At that point there are no roles, because they are set within firewall initialization and that's the reason, why I can not update user roles here.

So this can be considered as workaround, if roles have to be updated we notify listener to do so when it comes to its execution.